File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/promisehandlers/AddPartitioningHandler.scala
package edu.uci.ics.amber.engine.architecture.worker.promisehandlers

import com.twitter.util.Future
import edu.uci.ics.amber.engine.architecture.rpc.controlcommands.{
  AddPartitioningRequest,
  AsyncRPCContext
}
import edu.uci.ics.amber.engine.architecture.rpc.controlreturns.EmptyReturn
import edu.uci.ics.amber.engine.architecture.worker.DataProcessorRPCHandlerInitializer
import edu.uci.ics.amber.engine.architecture.worker.statistics.WorkerState.{PAUSED, READY, RUNNING}

trait AddPartitioningHandler {
  this: DataProcessorRPCHandlerInitializer =>

  override def addPartitioning(
      msg: AddPartitioningRequest,
      ctx: AsyncRPCContext
  ): Future[EmptyReturn] = {
    dp.stateManager.assertState(READY, RUNNING, PAUSED)
    dp.outputManager.addPartitionerWithPartitioning(msg.tag, msg.partitioning)
    EmptyReturn()
  }

}

File: core/amber/src/main/python/core/architecture/handlers/control/add_partitioning_handler.py
from core.architecture.handlers.control.control_handler_base import ControlHandler
from proto.edu.uci.ics.amber.engine.architecture.rpc import (
    EmptyReturn,
    AddPartitioningRequest,
)


class AddPartitioningHandler(ControlHandler):

    async def add_partitioning(self, req: AddPartitioningRequest) -> EmptyReturn:
        self.context.output_manager.add_partitioning(req.tag, req.partitioning)
        return EmptyReturn()
